### PR TITLE
Generate LibGit2 credential test expressions

### PR DESCRIPTION
**Pull request by omus**
Generate credential test expressions instead of writing them all by hand. Makes it easier to deal with refactoring.

Only possible since #23575 and #23641 were merged. Spin off of the work done in #20725.

### CODE REVIEW

**Comment by omus** [comment]
The Circle CI failure at first appears related but I think it may just be a coincidence as these tests work fine everywhere else and also worked before they were rebased.
```
Worker 4 terminated.
ERROR (unhandled task failure): EOFError: read end of file
...
libgit2: Error During Test
  Test threw an exception of type ProcessExitedException
  Expression: libgit2
  ProcessExitedException()
```

File: test/libgit2.jl [hunk 1]
```diff
@@ -1906,100 +1898,88 @@ mktempdir() do dir
...
 
             # Explicitly provided credential is correct
-            err, auth_attempts = challenge_prompt(valid_ex, [])
+            ex = gen_ex(valid_cred)
+            err, auth_attempts = challenge_prompt(ex, [])
             @test err == git_ok
             @test auth_attempts == 1
 
             # Explicitly provided credential is incorrect
-            # TODO: Unless the SSH agent is disabled we may get caught in an infinite loop
```
Comment (rofinn) [inline test/libgit2.jl]: Is this no longer an issue?

**Review by rofinn — APPROVED** [review]
I think this looks good, but I was a little unclear on a comment deletion. FWIW, it was definitely easier to follow the new code :)

File: test/libgit2.jl (hunk 1 above)
Comment (omus) [reply test/libgit2.jl]: Was fixed in #23641 when `use_ssh_agent=false` was set. I discovered the outdated comment while updating these tests:
https://github.com/JuliaLang/julia/blob/c172e9271d57cf2641def7fc362af2f1e71b456d/test/libgit2.jl#L1971

Comment (rofinn) [reply test/libgit2.jl]: Ahh, okay, thanks.

**Pull request merged**